### PR TITLE
[CPU] Enable barrier op upon gloo

### PR DESCRIPTION
**Pull request by bliu3650**
### PR types
New features

### PR changes
APIs

### Describe

- Background:
Barrier operation from specific communication library (e.g. HCCL for Ascend cluster) was not stable, and backup solution with CPU barrier is required;
CPU barrier op was developed before but not verified;
Considering performance across thousands of machine to synchronize, calling barrier function directly is better than executing it with program;

- Implementation:
Reuse previous barrier op implemented for CPU and add an initialization parallel environment flow for CPU only;
Both function-level call and op-level are supported;

- Unittest:
test_collective_cpu_barrier_with_gloo.py
截屏2021-08-18 下午6 10 24

### CODE REVIEW

File: python/paddle/distributed/parallel_with_gloo.py [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+
+    if rank_num < 2:
+        warnings.warn(
+            "Currently not a parallel execution environment, "
+            "`paddle.distributed.init_gloo_parallel_env` will not do anything.")
+        return
+
+    # init gloo context
+    if not with_gloo:
+        warnings.warn(
```
Comment (gongweibao) [inline python/paddle/distributed/parallel_with_gloo.py]: assert?

File: python/paddle/distributed/parallel_with_gloo.py [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+            "With gloo flag is not set. "
+            "`paddle.distributed.init_gloo_parallel_env` will not do anything.")
+        return
+    else:
```
Comment (gongweibao) [inline python/paddle/distributed/parallel_with_gloo.py]: 65行已经return了，这个地方就不用 else了？

File: python/paddle/fluid/tests/unittests/test_collective_cpu_barrier_with_gloo.py [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+        port = _free_port()
+        if port not in port_set:
+            port_set.add(port)
+            return port
+
+
+def barrier_op_func(id, rank_num, server_endpoint, out_dict, sleep_time):
+    paddle.distributed.init_gloo_parallel_env(id, rank_num, server_endpoint)
+    # 1st barrier
+    paddle.distributed.gloo_barrier()
```
Comment (gongweibao) [inline python/paddle/fluid/tests/unittests/test_collective_cpu_barrier_with_gloo.py]: barrier就好了，42行已经是说明是gloo了。

**Review by gongweibao — APPROVED** [review]
Coverage没过

**Review by gongweibao — CHANGES_REQUESTED** [review]
Coverage Failed

File: python/paddle/distributed/parallel_with_gloo.py (hunk 2 above)
Comment (bliu3650) [reply python/paddle/distributed/parallel_with_gloo.py]: 上面用了 assert，这个 else 分支也就不用了

File: python/paddle/distributed/parallel_with_gloo.py (hunk 1 above)
Comment (bliu3650) [reply python/paddle/distributed/parallel_with_gloo.py]: ok

File: python/paddle/fluid/tests/unittests/test_collective_cpu_barrier_with_gloo.py (hunk 3 above)
Comment (bliu3650) [reply python/paddle/fluid/tests/unittests/test_collective_cpu_barrier_with_gloo.py]: 在 __init__.py 文件里有个同名 barrier 函数，来自 collective.py （通过 append op 组网调用 barrier，也支持其他通信库）；这里写的 gloo_barrier 只针对 CPU，是不是需要区分一下？

**Review by gongweibao — DISMISSED** [review]
LGTM

**Review by PangHua — APPROVED** [review]

**Review by PangHua — DISMISSED** [review]

File: python/paddle/distributed/parallel_with_gloo.py [hunk 4]
```diff
@@ -0,0 +1,241 @@
...
+    from paddle.distributed.fleet.utils.http_server import KVServer
+    http_server = KVServer(int(port), size=size)
+    http_server.start()
+    wait_seconds = 3
+    while http_server_d.get("running", False) or not http_server.should_stop():
+        time.sleep(wait_seconds)
+    http_server.stop()
+
+
+def init_gloo_parallel_env(rank_id, rank_num, server_endpoint, with_gloo=True):
```
Comment (lanxianghit) [inline python/paddle/distributed/parallel_with_gloo.py]: 缺少文档预览

Comment (bliu3650) [reply python/paddle/distributed/parallel_with_gloo.py]: 是这个新增文件的整体预览，是吗？

**Comment by bliu3650** [comment]
新增APIs的文档预览：
init_gloo_parallel_env
barrier_func
release_gloo

**Review by lanxianghit — DISMISSED** [review]

**Review by XieYunshen — DISMISSED** [review]

**Review by PangHua — DISMISSED** [review]

**Review by gongweibao — DISMISSED** [review]
LGTM

**Comment by bliu3650** [comment]
新增 APIs （修改名称后）文档预览：
10 136 157 23_8090_documentation_docs_zh_api_paddle_distributed_gloo_init_parallel_env_cn html
10 136 157 23_8090_documentation_docs_zh_api_paddle_distributed_gloo_barrier_cn html
10 136 157 23_8090_documentation_docs_zh_api_paddle_distributed_gloo_release_cn html

**Review by PangHua — DISMISSED** [review]

**Review by XieYunshen — APPROVED** [review]
LGTM

**Review by PangHua — APPROVED** [review]

**Review by gongweibao — CHANGES_REQUESTED** [review]
LGTM

**Review by lanxianghit — APPROVED** [review]

**Review by zhiqiu — APPROVED** [review]
LGTM for init.py

**Review by gongweibao — APPROVED** [review]
LGTM

**Pull request merged**